docs/src/proposals/interest-bearing-tokens.md

```diff
@@ -0,0 +1,284 @@
+an ephemeral account with 8 bytes of data, passes it to the i-token program's
+`GetHoldingBalance` instruction, then reads the information back as a `u64`. At the
+end of the instruction, the account disappears.
+
+TODO What's the best approach to implement this?  For example, does an ephemeral
+account even need a public key?
+
+This solution avoids the need to pass additional scratch accounts to almost all
+programs that use tokens.
+
+TODO what limits do we consider for these? A certain number of accounts?
+A total amount of bytes? Regarding security, this breaks the previous
+check that a CPI must use a subset of accounts provided to the calling program.
+
+#### Dynamic sysvars
+
+The i-token program also requires the use of dynamic sysvars. For example,
+during the `GetHoldingBalance` instruction, the program needs to know the current
+time in order to properly convert from token amount to share amount.
+
+Without dynamic sysvars, we need to create a system to tell what sysvars are
+needed for different instructions on each program, which creates complications
+for all users.
+
+### RPC
+
+The validator's RPC server needs to properly handle new token program accounts
+for read-only instructions and pre / post balances.
+
+#### `getBalance` and `getSupply`
+
+Instead of deserializing SPL token accounts and returning the data, the
+RPC server runs read-only transactions on the i-token program.
+
+RPC simply uses the same flow as on-chain programs: create an ephemeral account
+then runs the read-only transaction onto the bank requested.
+
+TODO how do we set the cost of read-only transactions? Do we lower it specially
+for read-only RPC calls? The i-token program will probably be one of the more
+cost-intensive computations possible since it is performing present value
+discounting.
+
+#### `preTokenBalances` and `postTokenBalances`
+
+As mentioned earlier in the Token Program Conformance section, the i-token
+program caches balances in the `amount` field of the SPL token holding. This
+should only be used by RPC, and not by programs to get the balance of a non-SPL
+holding.
+
+### Associated Token Program
+
+The Associated Token Program needs to support all token programs seamlessly.
+Currently, it performs the following sequence:
+
+* `transfer` required lamports
+* `allocate` space
+* `assign` to the SPL token program
+* call `InitializeAccount` on SPL token
+
+This does not work for an opaque token program, because we do not know the size required
+from the outside. Conversely, if we allow a token program to take the lamports it wants,
+a malicious token program could take too much from users.
+
+To get around this, the Token Program interface has a new `CreateHolding`
+instruction, which only allocates space and assigns the account to the program.
+Once the space is allocated, the Associated Token Program can transfer the
+required lamports, and return an error if that number is too large. The process
```

Comment:
how much would be too large like over 10KB?

Reply:
I think for the case of NTFs (i.e. somewhat controlled environment), it might be desirable to have an option of raising the bar to carry around somewhat rich metadata of ownership transitions.

Reply:
I'm not sure to be honest -- 10KB seems pretty big for an account, but that would be a safe maximum.  Perhaps we start smaller and grow as needed?

In the case of an NFT, they could put most of the information in the mint so that token accounts stay small.  The mint is passed for all transfers so the program can do any necessary checks there.

Reply:
Might want to put the data into a separate account.  The NFT data might be large and we don't need to pass it around for transfers, etc.  Just put a reference to the NFT data account in the mint?

Reply:
I agree with @jackcmay NFTs data should go to a separate metadata account, the same way as in ERC721Metadata

Reply:
(this is out of scope of this proposal...)

@jackcmay @bartosz-lipinski Yeah, creating and separating account and referencing would work for normal NFT metadata. But, I was thinking about complex NFT which want to mutate the state of NFT metadata or want to hold past transfer history somehow. In that case, we need a writable flag for referenced mint account or additional NFT metadata, which isn't conforming to this i-token abi. So, I was thinking about putting those mutable NFT state in the already writable accounts (i.e. the dest token holding account with somewhat large buffer).

Then, ultimately I dreamed some fancy idea to conform the complex NFT to the i-token: https://github.com/solana-labs/solana/pull/15927#discussion_r595842064